src/storage/rect.rs

```diff
@@ -0,0 +1,632 @@
+use std::fmt::Debug;
+
+use glam::{IVec2, Vec2Swizzles};
+
+use crate::{Hex, HexLayout, OffsetHexMode, storage::HexStore};
+
+/// [`Vec`] Based storage for rectangular maps.
+///
+/// > See [this article](https://www.redblobgames.com/grids/hexagons/#map-storage)
+///
+/// [`RectMap`] is made for _rectangular_ large _dense_ maps, utilizing some
+/// tricks to map [`Hex`] coordinate to a positive 1D array.
+///
+/// It can be used only if:
+/// - The map is a rectanglar shape
+/// - The map is _dense_
+/// - No coordinate will be added or removed from the map
+///
+/// If your use case doesn't match all of the above, use a
+/// [`std::collections::HashMap`] instead
+///
+/// # Example
+/// ```rust
+/// use hexx::{
+///     storage::{HexStore, RectMap, RectMetadata, WrapStrategy},
+///     *,
+/// };
+///
+/// let layout = HexLayout::pointy()
+///     .with_hex_size(30.0)
+///     .with_origin(Vec2::ZERO);
+/// let rect_map = RectMetadata::default()
+///     .with_hex_layout(layout)
+///     .with_half_size(IVec2 { x: 8, y: 4 })
+///     .with_wrap_strategies([WrapStrategy::Cycle, WrapStrategy::Clamp])
+///     .build_default::<i32>();
+///
+/// assert_eq!(rect_map.get(Hex::new(0, 0)), Some(&0));
+/// ```
+///
+/// internally handle coordinate transform
+/// - `hex`
+/// - => `ij` offset coordinate
+/// - => `rc` 2D view of `Vec`
+/// - => `idx` `Vec` index
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(bevy_reflect::Reflect))]
+#[cfg_attr(
+    feature = "bevy_ecs",
+    derive(bevy_ecs::resource::Resource, bevy_ecs::component::Component)
+)]
+#[derive(Clone)]
+pub struct RectMap<T> {
+    inner: Vec<T>,
+    meta: RectMetadata,
+}
+
+/// Metadata struct for [`RectMap`]
+///
+/// # Example
+/// ```rust
+/// use hexx::{
+///     Hex, HexLayout, IVec2, OffsetHexMode,
+///     storage::{HexStore, RectMap, RectMetadata, WrapStrategy},
+/// };
+///
+/// let rect_hex_map = RectMetadata::default()
+///     .with_hex_layout(HexLayout::pointy().with_hex_size(1.0))
+///     .with_half_size(IVec2::new(8, 12))
+///     .with_offset_mode(OffsetHexMode::Odd)
+///     .with_wrap_strategies([WrapStrategy::Cycle, WrapStrategy::Clamp])
+///     .build_default::<i32>();
+/// let hex = Hex::new(0, 0);
+///
+/// assert_eq!(rect_hex_map.get(hex), Some(&0_i32));
+/// assert_eq!(rect_hex_map.wrapped_get(hex), &0_i32);
+/// ```
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(bevy_reflect::Reflect))]
+#[cfg_attr(
+    feature = "bevy_ecs",
+    derive(bevy_ecs::resource::Resource, bevy_ecs::component::Component)
+)]
+pub struct RectMetadata {
+    /// the hex layout of the map
+    hex_layout: HexLayout,
```

Comment:
```suggestion
    orientation: HexOrientation,
```

There doesn't seem to be any used field other than the orientation from the layout, so I don't think it makes sens to incorporate an entire hexagonal layout for a storage map.

The `HexLayout` exists to brigde between the hexagonal and the pixel space, when the storage is purely in the hexagonal space

Reply:
I think that the map should tie to the layout, since we can use the layout get the `Hex` coordinate from `Vec2`

if we don't include the layout inside the map, the user need to gather both the `RectMap` and `HexLayout` which might risk getting out od sync 

```rust
let hex = map.world_pos_to_hex(pos);
       // ^^^
       //   L > Deref RectMap => RectMetadata => HexLayout
```

this is also why I impl `Deref` for `RectMap=>RectMetada` and `RectMetadata=>HexLayout`, it help expose api without verbosity, I am still ironing out how to do that.

Reply:
A map store whether it is `RectMap` or `HexModMap` or the others is just a optimized hashmap `Hex` => `T`.

I understand it can be useful to have utility methods to convert pixel coordinates directly in the map but it's not the responsibility of a map storage to do that, it's only responsibility is mapping an `Hex` to a `T` value.

For the use case you have in mid there could be a resource with both a layout and a rectmap storage in it, and that resource could have useful methods to directly access the `T` value for any given pixel (`Vec2`) but it should not be part of the storage.
There is also no risk of those getting _out of sync_ since they do completely different things
- `HexLayout`: world/pixel coordinate (`Vec2`) <=> Hexagonal space (`Hex`)
- `RectMap`: Hexagonal space (`Hex`) => User defined value `T`